File: src/main/java/work/nekow/particledrawing/core/network/ParticleDestroyPayload.java
package work.nekow.particledrawing.core.network;

import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.resources.Identifier;
import org.jspecify.annotations.NonNull;

import java.util.*;

public record ParticleDestroyPayload(
    UUID[] particleIds,
    UUID groupId
) implements CustomPacketPayload {

    public static final Type<ParticleDestroyPayload> TYPE = new Type<>(
        Identifier.fromNamespaceAndPath("particledrawing", "particle_destroy")
    );

    public static final StreamCodec<FriendlyByteBuf, ParticleDestroyPayload> STREAM_CODEC =
        new StreamCodec<>() {
            @Override
            public @NonNull ParticleDestroyPayload decode(FriendlyByteBuf buf) {
                int count = buf.readVarInt();
                UUID[] ids = new UUID[count];
                for (int i = 0; i < count; i++) {
                    ids[i] = new UUID(buf.readLong(), buf.readLong());
                }
                boolean hasGroup = buf.readBoolean();
                UUID groupId = hasGroup ? new UUID(buf.readLong(), buf.readLong()) : null;
                return new ParticleDestroyPayload(ids, groupId);
            }

            @Override
            public void encode(FriendlyByteBuf buf, ParticleDestroyPayload payload) {
                buf.writeVarInt(payload.particleIds.length);
                for (UUID id : payload.particleIds) {
                    buf.writeLong(id.getMostSignificantBits());
                    buf.writeLong(id.getLeastSignificantBits());
                }
                buf.writeBoolean(payload.groupId != null);
                if (payload.groupId != null) {
                    buf.writeLong(payload.groupId.getMostSignificantBits());
                    buf.writeLong(payload.groupId.getLeastSignificantBits());
                }
            }
        };

    @Override
    public @NonNull Type<? extends CustomPacketPayload> type() {
        return TYPE;
    }

    public static ParticleDestroyPayload single(UUID particleId) {
        return new ParticleDestroyPayload(new UUID[]{particleId}, null);
    }

    public static ParticleDestroyPayload group(UUID groupId, Collection<UUID> memberIds) {
        return new ParticleDestroyPayload(memberIds.toArray(new UUID[0]), groupId);
    }

    public static ParticleDestroyPayload batch(Collection<UUID> ids) {
        return new ParticleDestroyPayload(ids.toArray(new UUID[0]), null);
    }
}

File: src/main/kotlin/work/nekow/particledrawing/core/network/ParticleDestroyPayload.kt
package work.nekow.particledrawing.core.network

import net.minecraft.network.FriendlyByteBuf
import net.minecraft.network.codec.StreamCodec
import net.minecraft.network.protocol.common.custom.CustomPacketPayload
import net.minecraft.resources.Identifier
import java.util.UUID

/**
 * 粒子销毁数据包，支持单个、批量和组的销毁。
 *
 * @param particleIds 要销毁的粒子 ID 数组
 * @param groupId 组 ID（组销毁时使用），可为 null
 */
data class ParticleDestroyPayload(
    val particleIds: Array<UUID>,
    val groupId: UUID?
) : CustomPacketPayload {

    companion object {
        @JvmField
        val TYPE = CustomPacketPayload.Type<ParticleDestroyPayload>(
            Identifier.fromNamespaceAndPath("particledrawing", "particle_destroy")
        )

        @JvmField
        val STREAM_CODEC: StreamCodec<FriendlyByteBuf, ParticleDestroyPayload> =
            object : StreamCodec<FriendlyByteBuf, ParticleDestroyPayload> {
                override fun decode(buf: FriendlyByteBuf): ParticleDestroyPayload {
                    val count = buf.readVarInt()
                    val ids = Array(count) { StreamCodecs.UUID_CODEC.decode(buf) }
                    val groupId = StreamCodecs.readNullableUUID(buf)
                    return ParticleDestroyPayload(ids, groupId)
                }

                override fun encode(buf: FriendlyByteBuf, payload: ParticleDestroyPayload) {
                    buf.writeVarInt(payload.particleIds.size)
                    for (id in payload.particleIds) StreamCodecs.UUID_CODEC.encode(buf, id)
                    StreamCodecs.writeNullableUUID(buf, payload.groupId)
                }
            }

        fun single(particleId: UUID): ParticleDestroyPayload {
            return ParticleDestroyPayload(arrayOf(particleId), null)
        }

        fun group(groupId: UUID, memberIds: Collection<UUID>): ParticleDestroyPayload {
            return ParticleDestroyPayload(memberIds.toTypedArray(), groupId)
        }

        @Suppress("unused")
        fun batch(ids: Collection<UUID>): ParticleDestroyPayload {
            return ParticleDestroyPayload(ids.toTypedArray(), null)
        }
    }

    override fun type(): CustomPacketPayload.Type<out CustomPacketPayload> = TYPE

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ParticleDestroyPayload) return false
        if (!particleIds.contentEquals(other.particleIds)) return false
        return groupId == other.groupId
    }

    override fun hashCode(): Int {
        var result = particleIds.contentHashCode()
        result = 31 * result + (groupId?.hashCode() ?: 0)
        return result
    }
}
